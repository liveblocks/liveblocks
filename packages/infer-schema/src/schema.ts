import { AST } from "@liveblocks/schema";

import type { InferredType } from "./inference";
import { isAtomic } from "./inference";
import {
  InferredObjectType,
  inferredObjectTypeToAst,
  isInferredObjectType,
  mergeInferredObjectTypes,
} from "./object";
import type { InferredTypeReference } from "./typeReference";
import { BidirectionalMap } from "./utils/bidirectionalMap";
import { invariant } from "./utils/invariant";
import { isNotUndefined } from "./utils/typeGuards";

type RootTypes = Set<InferredObjectType>;
type RootNames = BidirectionalMap<string, InferredObjectType>;

export type InferredSchema = {
  rootTypes: RootTypes;
  rootNames: RootNames;
};

// Deeply iterate over a inferred type, yielding the type and optionally the type reference
// object referencing it.
// If yieldCircularReferences is true, the same type can be yielded multiple times
// if another type circularly references it in order to yield all references to that type.
function* iterateInferredTypesDeep(
  root: InferredType,
  yieldCircularReferences = true
): Generator<[InferredType, InferredTypeReference | undefined]> {
  const visited = new Set<InferredType>();

  function* iterate(
    root: InferredType,
    referenceObj: InferredTypeReference | undefined = undefined
  ): Generator<[InferredType, InferredTypeReference | undefined]> {
    if (visited.has(root)) {
      if (yieldCircularReferences && referenceObj) {
        yield [root, referenceObj];
      }

      return;
    }

    visited.add(root);

    yield [root, referenceObj];
    switch (root.type) {
      case "Object":
        for (const field of Object.values(root.fields)) {
          yield* iterate(field.value, field);
        }
        return;
    }
  }

  yield* iterate(root);
}

function inferRootTypes(
  inferred: InferredObjectType
): Pick<InferredSchema, "rootTypes"> {
  const rootTypes: RootTypes = new Set();

  for (const [type] of iterateInferredTypesDeep(inferred)) {
    if (isInferredObjectType(type)) {
      rootTypes.add(type);
    }
  }

  return { rootTypes };
}
export function replaceRootType(
  schema: InferredSchema,
  oldType: InferredObjectType,
  newType: InferredObjectType
) {
  invariant(schema.rootTypes.has(oldType), "Old type not part of the schema");

  schema.rootTypes.delete(oldType);
  schema.rootTypes.add(newType);
  schema.rootNames.deleteValue(oldType);

  schema.rootTypes.forEach((rootType) => {
    for (const [, ref] of iterateInferredTypesDeep(rootType)) {
      if (ref?.value === oldType) {
        ref.value = newType;
      }
    }
  });
}

function assignNameOrMerge(schema: InferredSchema, type: InferredObjectType) {
  invariant(schema.rootTypes.has(type), "Root type not part of the schema");
  invariant(!schema.rootNames.hasValue(type), "Root type already has a name");

  const orderedNames = Object.entries(type.names)
    .sort(([, scoreA], [, scoreB]) => scoreB - scoreA)
    .map(([name]) => name);

  // TODO: Be smarter about when to merge types, and when to rename them.
  // Might also be worth a try to rename both types e.g. "Data" conflicts with "Data" ->
  // rename them to "UserData" and "ShapeData" instead of keeping "Data" and adding "UserData".
  for (const name of orderedNames) {
    const existingType = schema.rootNames.get(name);
    if (!existingType) {
      schema.rootNames.set(name, type);
      return;
    }

    const merged = mergeInferredObjectTypes(existingType, type, schema);
    if (merged) {
      schema.rootNames.set(name, merged);
      return;
    }
  }

  const baseName = orderedNames[0];
  invariant(isNotUndefined(baseName), "Expected at least one name");

  // Try Name2, Name3, etc. until we find a name that doesn't conflict.
  let n = 2;
  while (true) {
    const name = `${baseName}${n}`;
    if (!schema.rootNames.has(name)) {
      schema.rootNames.set(name, type);
      return;
    }
    n++;
  }
}

function getNextRootTypeToAssignName(schema: InferredSchema) {
  return Array.from(schema.rootTypes.values())
    .filter((type) => !schema.rootNames.hasValue(type))
    .sort((a, b) => {
      const aIsAtomic = isAtomic(a);
      const bIsAtomic = isAtomic(b);
      if (aIsAtomic && bIsAtomic) {
        return 0;
      }
      if (aIsAtomic || bIsAtomic) {
        return aIsAtomic ? -1 : 1;
      }

      // TODO: Assign names to root types with fewer aliases first?
      return (
        Math.max(...Object.values(a.names)) -
        Math.max(...Object.values(b.names))
      );
    })[0];
}

export function buildSchema(
  inferredStorage: InferredObjectType
): InferredSchema {
  const rootTypes = inferRootTypes(inferredStorage);
  const schema: InferredSchema = {
    ...rootTypes,
    rootNames: new BidirectionalMap(),
  };

  let toAssign = getNextRootTypeToAssignName(schema);
  while (toAssign) {
    assignNameOrMerge(schema, toAssign);
    toAssign = getNextRootTypeToAssignName(schema);
  }

  // TODO: Merge equal types?

  return schema;
}

export function inferredSchemaToAst(schema: InferredSchema): AST.Document {
  return AST.document(
    Array.from(schema.rootTypes.values()).map((rootType) =>
      inferredObjectTypeToAst(rootType, schema)
    )
  );
}
