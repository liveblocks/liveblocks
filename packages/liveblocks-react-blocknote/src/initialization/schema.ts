import type {
  BlockSchema,
  InlineContentSchema,
  StyleSchema,
} from "@blocknote/core";
import {
  BlockNoteSchema,
  createInlineContentSpecFromTipTapNode,
} from "@blocknote/core";
import { MentionNode } from "@liveblocks/react-tiptap";

const mentionSpec = createInlineContentSpecFromTipTapNode(MentionNode, {
  id: {
    default: "",
  },
  notificationId: {
    default: "",
  },
});
/**
 * Adds the Liveblocks Mention Node as inline content to the BlockNote schema
 *
 * This makes sure BlockNote knows about Liveblocks mentions and that you can read/write mentions via the BlockNote API
 */
export const withLiveblocksSchema = <
  B extends BlockSchema,
  I extends InlineContentSchema,
  S extends StyleSchema,
>(
  schema?: BlockNoteSchema<B, I, S>,
  liveblocksOptions: Partial<{ mentions: boolean }> = {}
): BlockNoteSchema<B, I, S> => {
  const optionalSchema = schema || BlockNoteSchema.create();
  if (!liveblocksOptions.mentions) {
    return optionalSchema as BlockNoteSchema<B, I, S>;
  }
  return BlockNoteSchema.create({
    blockSpecs: optionalSchema.blockSpecs,
    inlineContentSpecs: {
      ...optionalSchema.inlineContentSpecs,
      liveblocksMention: mentionSpec,
    },
    styleSpecs: optionalSchema.styleSpecs,
  }) as any as BlockNoteSchema<
    // typescript needs some help here
    B,
    I & {
      liveblocksMention: typeof mentionSpec.config;
    },
    S
  >;
};
