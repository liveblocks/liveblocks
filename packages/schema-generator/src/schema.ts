import { mergeInferredTypes, InferredType, isAtomic } from ".";
import { InferredObjectType, inferredObjectTypeToAst } from "./object";
import { InferredTypeReference } from "./typeReference";
import { invariant } from "./utils/invariant";
import { BidirectionalMap } from "./utils/bidirectionalMap";
import { AST } from "@liveblocks/schema";

type RootTypes = Set<InferredObjectType>;
type RootReferences = Map<InferredObjectType, Set<InferredTypeReference>>;
type RootNames = BidirectionalMap<string, InferredObjectType>;

export type InferredSchema = {
  rootTypes: RootTypes;
  rootTypeReferences: RootReferences;
  rootNames: RootNames;
};

// Deeply iterate over a inferred type, yielding the type and optionally the type reference
// object referencing it.
// If yieldCircularReferences is true, the same type can be yielded multiple times
// if another type circularly references it in order to yield all references to that type.
function* iterateInferredTypesDeep(
  root: InferredType,
  yieldCircularReferences = true
): Generator<[InferredType, InferredTypeReference | undefined]> {
  const visited = new Set<InferredType>();

  function* iterate(
    root: InferredType,
    referenceObj: InferredTypeReference | undefined = undefined
  ): Generator<[InferredType, InferredTypeReference | undefined]> {
    if (visited.has(root)) {
      if (yieldCircularReferences && referenceObj) {
        yield [root, referenceObj];
      }

      return;
    }

    visited.add(root);

    yield [root, referenceObj];
    switch (root.type) {
      case "Object":
        for (const field of root.fields.values()) {
          yield* iterate(field.value, field);
        }
        return;
    }
  }

  yield* iterate(root);
}

function inferRootTypes(
  inferred: InferredObjectType
): Pick<InferredSchema, "rootTypes" | "rootTypeReferences"> {
  const rootTypes: RootTypes = new Set();
  const rootTypeReferences: RootReferences = new Map();

  for (const [type, referenceObj] of iterateInferredTypesDeep(inferred)) {
    if (type.type === "Object") {
      rootTypes.add(type);

      if (referenceObj) {
        const references = rootTypeReferences.get(type) ?? new Set();
        references.add(referenceObj);
        rootTypeReferences.set(type, references);
      }
    }
  }

  return { rootTypes, rootTypeReferences };
}

function replaceRootType(
  schema: InferredSchema,
  oldType: InferredObjectType,
  newType: InferredObjectType
) {
  invariant(schema.rootTypes.has(oldType), "Old type not part of the schema");

  schema.rootTypes.delete(oldType);
  schema.rootTypes.add(newType);
  schema.rootNames.deleteValue(oldType);

  // Re-wire all references to the old type to the new type
  const oldTypeReferences = schema.rootTypeReferences.get(oldType);
  if (oldTypeReferences) {
    const newReferences = schema.rootTypeReferences.get(newType) ?? new Set();
    for (const reference of oldTypeReferences) {
      reference.value = newType;
      newReferences.add(reference);
    }
    schema.rootTypeReferences.set(newType, oldTypeReferences);
  }
}

function assignNameOrMerge(schema: InferredSchema, type: InferredObjectType) {
  invariant(schema.rootTypes.has(type), "Root type not part of the schema");
  invariant(!schema.rootNames.hasValue(type), "Root type already has a name");

  const orderedNames = Array.from(type.names.entries())
    .sort(([, scoreA], [, scoreB]) => scoreB - scoreA)
    .map(([name]) => name);

  // TODO: Be smarter about when to merge types, and when to rename them.
  // Might also be worth a try to rename both types e.g. "Data" conflicts with "Data" ->
  // rename them to "UserData" and "ShapeData" instead of keeping "Data" and adding "UserData".
  for (const name of orderedNames) {
    const existingType = schema.rootNames.get(name);
    if (!existingType) {
      schema.rootNames.set(name, type);
      return;
    }

    const combined = mergeInferredTypes(existingType, type) as
      | InferredObjectType
      | undefined;

    if (combined) {
      replaceRootType(schema, existingType, combined);
      replaceRootType(schema, type, combined);
      schema.rootNames.set(name, combined);
      return;
    }
  }

  // Try Name2, Name3, etc. until we find a name that doesn't conflict.
  let suffix = 2;
  while (true) {
    const name = `${orderedNames[0]}${suffix}`;
    if (!schema.rootNames.has(name)) {
      schema.rootNames.set(name, type);
      return;
    }

    suffix++;
  }
}

export function inferSchema(inferredStorage: InferredObjectType) {
  const rootTypes = inferRootTypes(inferredStorage);
  const schema: InferredSchema = {
    ...rootTypes,
    rootNames: new BidirectionalMap(),
  };

  const toAssign = Array.from(schema.rootTypes.values()).sort((a, b) => {
    const aIsAtomic = isAtomic(a);
    const bIsAtomic = isAtomic(b);
    if (aIsAtomic && bIsAtomic) {
      return 0;
    }
    if (aIsAtomic || bIsAtomic) {
      return aIsAtomic ? -1 : 1;
    }

    // TODO: Assign names to root types with fewer aliases first?
    return Math.max(...a.names.values()) - Math.max(...b.names.values());
  });

  toAssign.forEach((type) => assignNameOrMerge(schema, type));
  return schema;
}

export function inferredSchemaToAst(schema: InferredSchema): AST.Document {
  return {
    _kind: "Document",
    definitions: Array.from(schema.rootTypes.values()).map((rootType) =>
      inferredObjectTypeToAst(rootType, schema)
    ),
    range: [0, 0],
  };
}
