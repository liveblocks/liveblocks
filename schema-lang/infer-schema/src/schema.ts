import { AST } from "@liveblocks/schema";

import type { InferredType, MergeContext } from "./inference";
import { isAtomic } from "./inference";
import { orderedRootTypeNames } from "./naming";
import type { InferredObjectType } from "./object";
import {
  inferredObjectTypeToAst,
  isInferredObjectType,
  mergeInferredObjectTypes,
} from "./object";
import type { InferredTypeReference } from "./typeReference";
import { BidirectionalMap } from "./utils/bidirectionalMap";
import { invariant } from "./utils/invariant";
import { isNotUndefined } from "./utils/typeGuards";

type RootTypes = Set<InferredObjectType>;
type RootNames = BidirectionalMap<string, InferredObjectType>;

export type InferredSchema = {
  rootTypes: RootTypes;
  rootNames: RootNames;
};

// Deeply iterate over a inferred type, yielding the type and optionally the type reference
// object referencing it.
// If yieldCircularReferences is true, the same type can be yielded multiple times
// if another type circularly references it in order to yield all references to that type.
function* iterateInferredTypesDeep(
  root: InferredType,
  yieldCircularReferences = true
): Generator<[InferredType, InferredTypeReference | undefined]> {
  const visited = new Set<InferredType>();

  function* iterate(
    root: InferredType,
    referenceObj: InferredTypeReference | undefined = undefined
  ): Generator<[InferredType, InferredTypeReference | undefined]> {
    if (visited.has(root)) {
      if (yieldCircularReferences && referenceObj) {
        yield [root, referenceObj];
      }

      return;
    }

    visited.add(root);

    yield [root, referenceObj];
    switch (root.type) {
      case "Object":
        for (const field of Object.values(root.fields)) {
          yield* iterate(field.value, field);
        }
        return;
    }
  }

  yield* iterate(root);
}

function inferRootTypes(
  inferred: InferredObjectType
): Pick<InferredSchema, "rootTypes"> {
  const rootTypes: RootTypes = new Set();

  for (const [type] of iterateInferredTypesDeep(inferred)) {
    if (isInferredObjectType(type)) {
      rootTypes.add(type);
    }
  }

  return { rootTypes };
}

function applyRootTypeReplacements(
  schema: InferredSchema,
  replacements: Map<InferredObjectType, InferredObjectType>
): void {
  replacements.forEach((newType, oldType) => {
    invariant(schema.rootTypes.has(oldType), "Old type not found in schema");

    schema.rootTypes.delete(oldType);
    schema.rootTypes.add(newType);

    const currentName = schema.rootNames.getKey(oldType);
    schema.rootNames.deleteValue(oldType);

    // PERF: If the type already had his preferred name, we don't need to re-compute it
    const preferredName = orderedRootTypeNames(newType.names)[0];
    if (currentName !== undefined && currentName === preferredName) {
      schema.rootNames.set(currentName, newType);
    }
  });

  schema.rootTypes.forEach((rootType) => {
    for (const [, ref] of iterateInferredTypesDeep(rootType)) {
      if (!ref || !isInferredObjectType(ref.value)) {
        continue;
      }

      const replacement = replacements.get(ref.value);
      if (replacement) {
        ref.value = replacement;
      }
    }
  });
}

function mergeInferredRootTypes(
  schema: InferredSchema,
  a: InferredObjectType,
  b: InferredObjectType
): InferredObjectType | undefined {
  const mergeCtx: MergeContext = {
    mergeFns: new Map(),
    typeReplacements: new Map(),
  };

  const merged = mergeInferredObjectTypes(a, b, mergeCtx);
  if (!merged) {
    return;
  }

  const rootReplacements = new Map<InferredObjectType, InferredObjectType>();
  const references: Map<
    InferredObjectType,
    Set<InferredObjectType>
  > = new Map();

  // Abuse the face that maps iterate in insertion order
  mergeCtx.typeReplacements.forEach((newType, oldType) => {
    if (!isInferredObjectType(oldType) || !isInferredObjectType(newType)) {
      return;
    }

    if (schema.rootTypes.has(oldType)) {
      rootReplacements.set(oldType, newType);
    }

    const newTypeReferences = references.get(newType) ?? new Set();
    const oldTypeReferences = references.get(oldType);
    oldTypeReferences?.forEach((key) => {
      newTypeReferences.add(key);
      rootReplacements.set(key, newType);
    });

    references.set(newType, newTypeReferences);
  });

  applyRootTypeReplacements(schema, rootReplacements);
  return merged;
}

function assignNameOrMerge(schema: InferredSchema, type: InferredObjectType) {
  invariant(schema.rootTypes.has(type), "Root type not part of the schema");
  invariant(!schema.rootNames.hasValue(type), "Root type already has a name");

  const names = orderedRootTypeNames(type.names);

  // TODO: Be smarter about when to merge types, and when to rename them.
  // Might also be worth a try to rename both types e.g. "Data" conflicts with "Data" ->
  // rename them to "UserData" and "ShapeData" instead of keeping "Data" and adding "UserData".
  for (const name of names) {
    const existingType = schema.rootNames.get(name);
    if (!existingType) {
      schema.rootNames.set(name, type);
      return;
    }

    const merged = mergeInferredRootTypes(schema, existingType, type);
    if (merged) {
      schema.rootNames.set(name, merged!);
      return;
    }
  }

  const preferredName = names[0];
  invariant(isNotUndefined(preferredName), "Expected at least one name");

  // Try Name2, Name3, etc. until we find a name that doesn't conflict.
  let n = 2;
  while (true) {
    const name = `${preferredName}${n}`;
    if (!schema.rootNames.has(name)) {
      schema.rootNames.set(name, type);
      return;
    }
    n++;
  }
}

function getNextRootTypeToAssignName(schema: InferredSchema) {
  return Array.from(schema.rootTypes.values())
    .filter((type) => !schema.rootNames.hasValue(type))
    .sort((a, b) => {
      const aIsAtomic = isAtomic(a);
      const bIsAtomic = isAtomic(b);
      if (aIsAtomic && bIsAtomic) {
        return 0;
      }
      if (aIsAtomic || bIsAtomic) {
        return aIsAtomic ? -1 : 1;
      }

      // TODO: Assign names to root types with fewer aliases first?
      return (
        Math.max(...Object.values(a.names)) -
        Math.max(...Object.values(b.names))
      );
    })[0];
}

export function buildSchema(
  inferredStorage: InferredObjectType
): InferredSchema {
  const rootTypes = inferRootTypes(inferredStorage);
  const schema: InferredSchema = {
    ...rootTypes,
    rootNames: new BidirectionalMap(),
  };

  let toAssign = getNextRootTypeToAssignName(schema);
  while (toAssign) {
    assignNameOrMerge(schema, toAssign);
    toAssign = getNextRootTypeToAssignName(schema);
  }

  // TODO: Merge equal types?

  return schema;
}

export function inferredSchemaToAst(schema: InferredSchema): AST.Document {
  return AST.document(
    Array.from(schema.rootTypes.values()).map((rootType) =>
      inferredObjectTypeToAst(rootType, schema)
    )
  );
}
