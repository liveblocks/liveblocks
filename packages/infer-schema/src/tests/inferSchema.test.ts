import { parse } from "@liveblocks/schema";
import fc from "fast-check";

import type { PlainLsonObject } from "..";
import { inferSchema } from "..";
import { plainLsonArbitraries } from "./arbitraries";
import {
  BASIC_LIVE_OBJECT,
  BASIC_MERGE,
  BASIC_UNMERGEABLE,
  BRACKET_KEY,
  CIRCULAR_MERGE,
  EMPTY,
  EMPTY_KEY,
  KEY_WITH_WHITESPACE,
  RESERVED_KEY,
} from "./testData";

describe("inferSchema", () => {
  const testCases = {
    EMPTY,
    BASIC_MERGE,
    BASIC_UNMERGEABLE,
    BASIC_LIVE_OBJECT,
    RESERVED_KEY,
    BRACKET_KEY,
    KEY_WITH_WHITESPACE,
    EMPTY_KEY,
    CIRCULAR_MERGE,
  };

  Object.entries(testCases).forEach(([name, storageData]) => {
    it(`correctly infers the "${name}" schema`, () => {
      expect(inferSchema(storageData)).toMatchSnapshot();
    });
  });

  it("always generates a valid schema that matches the plain lson or includes a fixme comment", () => {
    fc.assert(
      fc.property(
        plainLsonArbitraries.object,

        (storageData) => {
          const schemaText = inferSchema(storageData as PlainLsonObject);
          if (!schemaText.includes("# FIXME: ")) {
            expect(() => parse(schemaText)).not.toThrow();
          }
        }
      )
    );
  });
});
