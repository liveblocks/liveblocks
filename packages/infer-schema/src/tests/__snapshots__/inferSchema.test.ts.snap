// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inferSchema circular 1`] = `
"type Storage {
  a: LiveObject<A>
}

type B {
  a: LiveObject<A>
}

type A {
  a?: LiveObject<A>
  b?: LiveObject<B>
}"
`;

exports[`inferSchema correctly infers the "BASIC_LIVE_OBJECT" schema 1`] = `
"type Storage {
  fills: LiveObject<Fill>
  strokes: LiveObject<Stroke>
}

type Fill {
  mesh: String
  sole: String
  stripes: String
  laces: String
}

type Stroke {
  mesh: String
  sole: String
  laces: String
}"
`;

exports[`inferSchema correctly infers the "BASIC_MERGE" schema 1`] = `
"type Storage {
  fills: LiveObject<Fill>
  strokes: LiveObject<Stroke>
}

type Fill {
  position: Position
  sole: String
  stripes: String
  laces: String
}

type Stroke {
  position: Position
  sole: String
  laces: String
}

type Position {
  x: Float
  y: Float
}"
`;

exports[`inferSchema correctly infers the "BASIC_UNMERGEABLE" schema 1`] = `
"type Storage {
  fills: LiveObject<Fill>
  strokes: LiveObject<Stroke>
}

type Fill {
  mesh: Mesh
  sole: String
  stripes: String
  laces: String
}

type Mesh {
  a: Int
}

type Stroke {
  mesh: StrokeMesh
  sole: String
  laces: String
}

type StrokeMesh {
  a: String
}"
`;

exports[`inferSchema correctly infers the "BRACKET_KEY" schema 1`] = `
"type Storage {
  _____: LiveObject<D> # FIXME: Field name 'd{' is illegal, can only contain alphanumeric characters and underscores
  _____2: LiveObject<D> # FIXME: Field name 'd]' is illegal, can only contain alphanumeric characters and underscores
}

type D {
}"
`;

exports[`inferSchema correctly infers the "EMPTY" schema 1`] = `
"type Storage {
}"
`;

exports[`inferSchema correctly infers the "EMPTY_KEY" schema 1`] = `
"type Storage {
  _____: LiveObject<Value> # FIXME: Field name '' is illegal, must start with a letter
}

type Value {
}"
`;

exports[`inferSchema correctly infers the "KEY_WITH_WHITESPACE" schema 1`] = `
"type Storage {
  _____: LiveObject<TestKey> # FIXME: Field name 'test key' is illegal, can only contain alphanumeric characters and underscores
}

type TestKey {
}"
`;

exports[`inferSchema correctly infers the "RESERVED_KEY" schema 1`] = `
"type Storage {
  _____: LiveObject<LiveblockType> # FIXME: Field name 'liveblocksType' is illegal, cannot be a reserved name
}

type LiveblockType {
}"
`;
